## Backtesting result of 007 pytorch model

In [8]:
from backtesting import Backtest, Strategy
import pandas as pd

hist = pd.read_csv("hist5.csv")
# print(hist.shape)

df = pd.read_csv("df007.csv")

df['u_sig'] = hist['0'] > 0.3
df['d_sig'] = hist['0'] < 0

# set index
df['Datetime'] = pd.DatetimeIndex(df['Datetime'])
df = df.set_index('Datetime')

df

Open         High          Low        Close  \
Datetime                                                                        
2023-12-05 13:00:00-05:00  4565.129883  4565.790039  4562.390137  4563.959961   
2023-12-05 13:05:00-05:00  4563.939941  4565.330078  4561.660156  4563.620117   
2023-12-05 13:10:00-05:00  4563.720215  4564.930176  4562.950195  4564.379883   
2023-12-05 13:15:00-05:00  4564.390137  4566.930176  4564.120117  4566.569824   
2023-12-05 13:20:00-05:00  4566.649902  4568.399902  4566.299805  4567.970215   
...                                ...          ...          ...          ...   
2024-02-16 15:30:00-05:00  5016.120117  5017.200195  5013.790039  5016.020020   
2024-02-16 15:35:00-05:00  5016.140137  5016.500000  5011.629883  5012.060059   
2024-02-16 15:40:00-05:00  5012.089844  5012.089844  5005.359863  5006.509766   
2024-02-16 15:45:00-05:00  5006.459961  5008.359863  5000.640137  5000.839844   
2024-02-16 15:50:00-05:00  5000.870117  5007.560059  5000.520020  5005.089844   

                             Adj Close     Volume       atr  u_sig  d_sig  
Datetime                                                                   
2023-12-05 13:00:00-05:00  4563.959961   35681144  4.207037  False   True  
2023-12-05 13:05:00-05:00  4563.620117   38296882  4.168672  False   True  
2023-12-05 13:10:00-05:00  4564.379883   33246920  4.012337  False   True  
2023-12-05 13:15:00-05:00  4566.569824   33388510  3.926460  False   True  
2023-12-05 13:20:00-05:00  4567.970215   28071558  3.796005  False  False  
...                                ...        ...       ...    ...    ...  
2024-02-16 15:30:00-05:00  5016.020020   53007682  3.659394  False  False  
2024-02-16 15:35:00-05:00  5012.060059   54369388  3.745874   True  False  
2024-02-16 15:40:00-05:00  5006.509766   75685062  3.959025  False  False  
2024-02-16 15:45:00-05:00  5000.839844   92109550  4.227646  False   True  
2024-02-16 15:50:00-05:00  5005.089844  145210136  4.428532  False   True  

[3935 rows x 9 columns]

In [9]:
class St(Strategy):
    def init(self):
        pass
        
    def next(self):
        if self.data['u_sig'][-1]: 
            self.buy()

        if self.position and self.data['d_sig'][-1]:
            self.position.close()

In [10]:
bt = Backtest(
    df,
    St,
    trade_on_close=True,
    exclusive_orders=True 
)

stats = bt.run()

stats

Start                     2023-12-05 13:00...
End                       2024-02-16 15:50...
Duration                     73 days 02:50:00
Exposure Time [%]                   41.219822
Equity Final [$]                 10913.408203
Equity Peak [$]                  10941.808594
Return [%]                           9.134082
Buy & Hold Return [%]                9.665507
Return (Ann.) [%]                   54.699259
Volatility (Ann.) [%]                8.194943
Sharpe Ratio                         6.674757
Sortino Ratio                       20.634319
Calmar Ratio                        40.029702
Max. Drawdown [%]                   -1.366467
Avg. Drawdown [%]                    -0.11422
Max. Drawdown Duration       11 days 18:25:00
Avg. Drawdown Duration        0 days 14:00:00
# Trades                                  778
Win Rate [%]                        58.997429
Best Trade [%]                       0.844991
Worst Trade [%]                     -0.466697
Avg. Trade [%]                    

In [11]:
bt.plot()

/Users/kimfaicheang/workspace/snowball/venv/lib/python3.10/site-packages/backtesting/_plotting.py:455: FutureWarning: 'H' is deprecated and will be removed in a future version, please use 'h' instead.
  df2 = (df.assign(_width=1).set_index('datetime')


Row(id='2788', ...)

Profit 109%，USD 913， 看起來好像還行，實際上錯的離譜，因為沒有算 commission，原本以為 0.002，千分之二的 commission 可以忽略不計，然而小數怕長計，SPX 一份 share USD 4500-5000， 也就是說每次買賣的 commission 已經來到 USD 10 上下，上面交易了778次，實際上付出了USD
 7千多的 commission..

In [12]:
bt = Backtest(
    df,
    St,
    commission=.002,
    trade_on_close=True,
    exclusive_orders=True 
)

stats = bt.run()

stats

Start                     2023-12-05 13:00...
End                       2024-02-16 15:50...
Duration                     73 days 02:50:00
Exposure Time [%]                   27.750953
Equity Final [$]                  4862.027807
Equity Peak [$]                       10000.0
Return [%]                         -51.379722
Buy & Hold Return [%]                9.665507
Return (Ann.) [%]                  -97.047543
Volatility (Ann.) [%]                0.604312
Sharpe Ratio                              0.0
Sortino Ratio                             0.0
Calmar Ratio                              0.0
Max. Drawdown [%]                  -51.381123
Avg. Drawdown [%]                  -51.381123
Max. Drawdown Duration       73 days 00:50:00
Avg. Drawdown Duration       73 days 00:50:00
# Trades                                  529
Win Rate [%]                         1.890359
Best Trade [%]                       0.643704
Worst Trade [%]                     -0.665367
Avg. Trade [%]                    

In [13]:
bt.plot()

/Users/kimfaicheang/workspace/snowball/venv/lib/python3.10/site-packages/backtesting/_plotting.py:455: FutureWarning: 'H' is deprecated and will be removed in a future version, please use 'h' instead.
  df2 = (df.assign(_width=1).set_index('datetime')


Row(id='3478', ...)

虧到底褲都沒了

不過剛才去試了下，一份 US500 的點差是 0.4，即 USD 0.4，相對於 4千多的原價，差不多是 0.0001 的水平

In [29]:
class St2(Strategy):
    def init(self):
        pass
        
    def next(self):
        if self.data['u_sig'][-1] and self.data['atr'] > 2: 
            self.buy()

        if self.position and self.data['d_sig'][-1]:
            self.position.close()

In [30]:
bt = Backtest(
    df,
    St2,
    commission=.0001,
    trade_on_close=True,
    exclusive_orders=True 
)

stats = bt.run()

stats

Start                     2023-12-05 13:00...
End                       2024-02-16 15:50...
Duration                     73 days 02:50:00
Exposure Time [%]                   39.034307
Equity Final [$]                 10208.061032
Equity Peak [$]                  10242.491805
Return [%]                            2.08061
Buy & Hold Return [%]                9.665507
Return (Ann.) [%]                   11.402363
Volatility (Ann.) [%]                6.096997
Sharpe Ratio                         1.870161
Sortino Ratio                        3.046429
Calmar Ratio                         5.814102
Max. Drawdown [%]                   -1.961156
Avg. Drawdown [%]                    -0.22314
Max. Drawdown Duration       36 days 20:45:00
Avg. Drawdown Duration        2 days 03:02:00
# Trades                                  731
Win Rate [%]                        51.709986
Best Trade [%]                       0.834908
Worst Trade [%]                      -0.47665
Avg. Trade [%]                    

In [31]:
bt.plot()

/Users/kimfaicheang/workspace/snowball/venv/lib/python3.10/site-packages/backtesting/_plotting.py:455: FutureWarning: 'H' is deprecated and will be removed in a future version, please use 'h' instead.
  df2 = (df.assign(_width=1).set_index('datetime')


Row(id='5548', ...)

聊勝於無